proposals/eligibility_validator.go

```diff
@@ -80,19 +83,29 @@ func (v *Validator) CheckEligibility(ctx context.Context, ballot *types.Ballot)
 	}
 
 	// todo: optimize by using reference to active set size and cache active set size to not load all atxsIDs from db
+	var owned *types.ActivationTxHeader
 	for _, atxID := range activeSets {
 		atx, err := v.cdb.GetAtxHeader(atxID)
 		if err != nil {
 			return false, fmt.Errorf("get ATX header %v: %w", atxID, err)
 		}
 		totalWeight += atx.GetWeight()
+		if atxID == ballot.AtxID {
+			owned = atx
+		}
 	}
-
-	atx, err := v.getBallotATX(ctx, ballot)
-	if err != nil {
-		return false, fmt.Errorf("get ballot ATX header %v: %w", ballot.AtxID, err)
+	if owned == nil {
+		return false, fmt.Errorf("atx %v from ballot %v (refballot %v) is not included into the active set", ballot.AtxID, ballot.ID(), refBallot.ID())
+	}
+	if targetEpoch := owned.TargetEpoch(); targetEpoch != epoch {
+		return false, fmt.Errorf("%w: ATX target epoch (%v), ballot publication epoch (%v)",
+			errTargetEpochMismatch, targetEpoch, epoch)
+	}
+	if pub := ballot.SmesherID(); !bytes.Equal(owned.NodeID.Bytes(), pub.Bytes()) {
```

Comment:
you can do "==" with types.NodeID (types.Hash32)

Reply:
copied this